### PR TITLE
AB#1176 Template sanity check

### PR DESCRIPTION
**Pull request by daniel-weisse**
### Proposed changes
- Perform a check on the templates of all file and env declarations during `cc.SetManifest`  to make sure:
  1. All Secrets that are to be injected actually exist
  2. Only valid encodings are used (For example: no symmetric key is assigned PEM encoding)
  3. No raw encoding for secrets injected as environment variables, except for secrets of type `plain`
      *  If a `plain` secret contains NULL bytes, the deployment of the Marble fails. Since `plain` secrets are user-defined, this can be fixed at run time.

### Additional info
I have also made some errors produce by invalid templates more clear. Instead of receiving `"invalid secret type"` when a template references a non existent secret, we now get `"secret does not exist"`.
Combined with the extra information of what Marble and what file/env contains the invalid declaration, this should help users to find their mistake more easily.

### CODE REVIEW

File: coordinator/manifest/manifest.go [hunk 1]
```diff
@@ -457,14 +461,37 @@ func EncodeSecretDataToBase64(data interface{}) (string, error) {
...
 }
 
-// ManifestTemplateFuncMap defines the functions which can be specified for secrets in the in go template format
-var ManifestTemplateFuncMap = template.FuncMap{
+// EncodeEnvSecretDataToRaw encodes only secrets of type plain to raw byte string
+func EncodeEnvSecretDataToRaw(data interface{}) (string, error) {
+	switch secret := data.(type) {
+	case Secret:
+		if secret.Type != "plain" {
+			return "", errors.New("only secrets of type plain are allowed to use raw encoding for environment variables")
```
Comment (Nirusu) [inline coordinator/manifest/manifest.go:469]: I would still check if these secrets do contain unexpected NULL bytes and also throw an error in this case.

File: coordinator/manifest/manifest.go [hunk 2]
```diff
@@ -457,14 +461,37 @@ func EncodeSecretDataToBase64(data interface{}) (string, error) {
 	return base64.StdEncoding.EncodeToString([]byte(raw)), nil
 }
 
-// ManifestTemplateFuncMap defines the functions which can be specified for secrets in the in go template format
-var ManifestTemplateFuncMap = template.FuncMap{
+// EncodeEnvSecretDataToRaw encodes only secrets of type plain to raw byte string
+func EncodeEnvSecretDataToRaw(data interface{}) (string, error) {
```
Comment (Nirusu) [inline coordinator/manifest/manifest.go:465]: I don't quite like to re-use the type `raw` here but impose limits onto it when we don't do it for Files. This might be unexpected.

**Review by Nirusu — COMMENTED** [review]
Quick review, have not tried to test (or break) it yet.

File: coordinator/manifest/manifest.go (hunk 1 above)
Comment (daniel-weisse) [reply coordinator/manifest/manifest.go:469]: I have added a general check for NULL bytes during `SetManifest` and moved the check at runtime during a Marble activation so it catches all NULL bytes.

I think this is better than to have the check in the template function itself, since NULL bytes can also occur if we dont use templates at all. For example through `base64` encoded env variables.

Comment (Nirusu) [reply coordinator/manifest/manifest.go:469]: I think checks in `SetManifest` and `WriteSecrets` for user-defined secrets are sufficient, does not necessarily be in the template function if these are the only two functions deploying new secrets.

An extra error during Marble activation is fine I guess, though not necessary cause NULL bytes are actually catched by `os.Setenv`. It's just the error is ugly and non-explanatory by default (it's just "invalid argument").

File: coordinator/manifest/manifest.go (hunk 2 above)
Comment (m1ghtym0) [reply coordinator/manifest/manifest.go:465]: What about a more explicit type like `env` or smth.?

File: coordinator/manifest/manifest.go (hunk 1 above)
Comment (m1ghtym0) [reply coordinator/manifest/manifest.go:469]: > I think checks in `SetManifest` and `WriteSecrets` for user-defined secrets are sufficient, does not necessarily be in the template function if these are the only two functions deploying new secrets.
> 
> An extra error during Marble activation is fine I guess, though not necessary cause NULL bytes are actually catched by `os.Setenv`. It's just the error is ugly and non-explanatory by default (it's just "invalid argument").

I don't quite understand what you're trying to say here, could you rephrase?:-)

Comment (Nirusu) [reply coordinator/manifest/manifest.go:469]: Basically check for (potential) NULL bytes whenever:

- We set a new manifest
- Someone updates an user-defined secret (which is piped into the environment)

Which should disallow any NULL bytes ideally. If for some other reasons there are still NULL bytes inserted during Marble activation, it will fail anyway, so we do not need any additional checks during runtime in my opinion, except if we want to return a user-friendly error to the user.

File: coordinator/manifest/manifest.go (hunk 2 above)
Comment (Nirusu) [reply coordinator/manifest/manifest.go:465]: The problem with `env` is that when you can also inject stuff like base64, what is `env` supposed to represent? I would just go with `(c)string` maybe.

File: coordinator/manifest/manifest.go (hunk 1 above)
Comment (m1ghtym0) [reply coordinator/manifest/manifest.go:469]: Thanks! Agreed

File: coordinator/manifest/manifest.go (hunk 2 above)
Comment (m1ghtym0) [reply coordinator/manifest/manifest.go:465]: Ok, let's move on here. I'm fine with `string`, if Daniel agrees let's just use it:-)

Comment (daniel-weisse) [reply coordinator/manifest/manifest.go:465]: string is OK.
As long as we make it clear in the documentation which tags do what, there shouldn't be any problems

File: coordinator/manifest/manifest.go (hunk 1 above)
Comment (daniel-weisse) [reply coordinator/manifest/manifest.go]: We now have checking for NULL bytes on manifest set, and every time a Marble activation occurs

For manifest set we do a general check on the env variable string to catch any NULL bytes explicitly written by the person setting the manifest (e.g. a base64 encoded string which contains NULL bytes).

For a Marble activation, this check occurs directly in the template function: If the encoding is of type `string` (formerly `raw`), only data without NULL bytes result in a starting Marble. Otherwise an error is returned notifying the user of the NULL byte.

Comment (Nirusu) [reply coordinator/manifest/manifest.go]: Okay, not I get the runtime checks. But should we not add another check for `WriteSecrets` upon upload of an user-defined secret?

Comment (Nirusu) [reply coordinator/manifest/manifest.go]: I would just re-run the template dry-run. If it's just defined as a file, it should be fine, if it's defined as env, it should fail. Similar to as when you would try to launch such a Marble.

Comment (daniel-weisse) [reply coordinator/manifest/manifest.go]: Added a template dry run in `WriteSecrets` :)

**Review by Nirusu — APPROVED** [review]
